core/src/main/scala/kafka/log/LogCleanerManager.scala

```diff
@@ -179,11 +179,15 @@ private[log] class LogCleanerManager(val logDirs: Seq[File],
         case (topicPartition, log) => // create a LogToClean instance for each
           try {
             val lastCleanOffset = lastClean.get(topicPartition)
-            val (firstDirtyOffset, firstUncleanableDirtyOffset) = cleanableOffsets(log, lastCleanOffset, now)
-            val compactionDelayMs = maxCompactionDelay(log, firstDirtyOffset, now)
+            val logStartOffset = log.logStartOffset
+            val offsetsToClean = cleanableOffsets(log, lastCleanOffset, now)
+            // update checkpoint for logs with invalid checkpointed offsets
+            if (offsetsToClean.needUpdateCheckpoint)
+              updateCheckpoints(log.dir.getParentFile(), Option(topicPartition, logStartOffset))
```

Comment:
It's better to use `offsetsToClean. firstDirtyOffset` instead of `logStartOffset` here, because that would remove the assumption in the code that `offsetsToClean. firstDirtyOffset` would always be log start offset when `needUpdateCheckpoint` is true.